### ISSUE TITLE
[Question] Is causal masking of recomputed tokens not correct?

### ISSUE THREAD

**Issue by Ray-Eldath** [LMCache/LMCache#1152]
I've noticed that when blending happens, `qkv` for recomputed and unmatched tokens are calculated freshly, but `k` and `v` gets blended into cached `kv`, while `q` gets separated out and feed into `self.attn` directly:

1. before blending, `qkv` are all for fresh tokens so their lengths are the same:
   <img width="230" height="233" alt="Image" src="https://github.com/user-attachments/assets/90d64799-fdb6-4c3d-ba55-d85399546349" />
2. after blending, `k` and `v` gets expanded to **the sequence length**, which is the total of "fresh tokens blended into cached tokens", while `q` still maintains the same:
   <img width="185" height="153" alt="Image" src="https://github.com/user-attachments/assets/ee13d1b8-40c6-43c1-9eae-da0679275fec" />

after blending, the full `kv` and fresh `q` gets fed into `self.attn`, which means length of `q` isn't the same as length of `k` and `v`. The hooked flash attention implementation (`flash_attn_varlen_func`) has a comment regarding *the causal masking* when `len(q) != len(kv)`:
> If causal=True, the causal mask is aligned to the bottom right corner of the attention matrix.
> For example, if seqlen_q = 2 and seqlen_k = 5, the causal mask (1 = keep, 0 = masked out) is:
>     1 1 1 1 0
>     1 1 1 1 1

which says, if `len(q) != len(kv)`, `q` tokens will attend to **all previous `kv`-s**, irrelevant of the position of `q`. So, for recomputed tokens, they will attend to tokens not only tokens *before* their original position, **but also tokens after their original position**, thus rendering the causal masking incorrect.

---

For example, suppose I have a sequence of 5 tokens, out of them tokens at position `[2, 4]` needs to be recomputed (fresh tokens). After blending, the `kv` will be length 5, and `q` will be length `2` (length of fresh tokens). According to the aforementioned comments, the causal masking for them will be:
```
1 2 3 4 5 6 7
------------
1 1 1 1 1 1 0    (attends to all 5 tokens)
1 1 1 1 1 1 1    (attends to all 5 tokens)
```
while *the correct* causal masking should be:
```
1 2 3 4 5 6 7
------------
1 0 0 0 0 1 0    (only attends to tokens before 2)
1 1 1 0 0 0 1    (only attends to tokens before 4)
```

---

Is my understanding correct? Can the model still maintains decent accuracy while the causal masking of recomputed tokens isn't correct?

Thank you very much.

**Comment by nechamab1** [comment]
@Ray-Eldath Yes, you're correct.
I asked this question in the Lmcache Slack, and here's the answer @YaoJiayi gave:

> The attention mask is a tricky thing. Using casual mask has some accuracy degradation (the accuracy drop is acceptible when the chunk numbers are small). If we want to use the accurate mask, we need either (1) customize the attention kernels or (2) leverage existing sparse attention libraries like flashinfer
I have done some initial benchmark on (2), but I haven't got time to push it to the upstream
This is on my roadmap, but feel free to contribute to this together!

**Comment by Ray-Eldath** [comment]
Now I see. Thank you very much! 😄

**Comment by yuang-lee** [comment]
Hi @Ray-Eldath @nechamab1 @YaoJiayi
Do you know whether this causal masking mismatch is resolved by the latest code in the LMCache project?
I can't find any information about this.
Should I look into the [lmcache-vllm](https://github.com/LMCache/lmcache-vllm)? but the last commit there is around two years ago, which is earlier than this issue, so I assume the mismatch of masking still exists?

**Comment by zihanlin-ai** [comment]
@yuang-lee Still present on `dev`: `LMCFlashAttnBackend.forward_contiguous` passes `causal=True` with `len(q) < len(k)` (`lmcache/v1/compute/attention/flash_attn.py:91`), so the bottom-right alignment described above still applies. The position-correct mask exists only in the opt-in sparse backends (`enable_sparse=True` → `update_from_top_indices`).

Cost, measured by replaying both masks through a reimplementation of the blend path on Qwen3-4B (256-token chunks, `blend_check_layers=[1]`, fp32 attention) against a full prefill — relative error of the last-position logits:

| chunks | ratio | position-correct | bottom-right |
| -----: | ----: | ---------------: | -----------: |
| 3 | 0.25 | 0.093 | 0.174 |
| 3 | 0.50 | 0.043 | 0.074 |
| 6 | 0.25 | 0.120 | 0.235 |
| 6 | 0.50 | 0.059 | 0.199 |

The two coincide at ratio 1, where `len(q) == len(k)`. The gap grows with chunk count, as @YaoJiayi expected. The practical effect is that recomputation buys much less than it should: at 6 chunks, raising the ratio from 0.05 to 0.5 cuts the error 4x with the correct mask but only 1.3x with bottom-right.

@YaoJiayi @ApostaC would you take a PR making the sparse backend the default for blending instead of opt-in?
